### ISSUE TITLE
Faster 1st order methods

### ISSUE THREAD

**Issue by SamuelGabriel** [f-dangel/backpack#94]
I really like your work on BackPack it enabled me to restructure my current research pretty massively. 

I have a question, though: Is it possible to speed up the batchl2norm for example? I implemented a similar extension that I use, and also rely on `convUtils.get_weight_gradient_factors` for convolutions. For my use cases this is really slow though. On a Wide ResNet 28x10, I have steps that take roughly 3 times longer if I use either batchl2norm or my own extension. My guess would be that this is due to the simplification call to `convUtils.get_weight_gradient_factors`.  Is that right?

I guess it would be a lot of work, but is it in general possible to implement batchl2norm, as fast as the sum of the gradients performed during backward for convolutions?

**Comment by f-dangel** [comment]
Hi, thanks for your feedback. 

Could you provide information about how your custom extension compares to approach 1 below, i.e. computing the individual gradients via `backpack.extensions.BatchGrad`, then taking the ℓ₂ norm manually?

If this explains the slowdown for your specific setting, we could analyze the trade-off between the different approaches to compute individual gradient ℓ₂ norms and maybe come up with a clever way to dynamically switch between them.

----

**Background:** I know about two ways to compute the individual gradient ℓ₂ norms for convolutions. The [code](https://github.com/f-dangel/backpack/blob/c3171ba9c144ec7b8eed781cd7126896cb9a3481/backpack/extensions/firstorder/batch_l2_grad/conv2d.py#L15-L19) is currently using variant 2:
1. Compute individual gradients, take ℓ₂ norm. For that, the individual gradients have to be built up explicitly in memory, which may be expensive. 
2. Interpret convolution as matrix multiplication between the matrix-reshaped kernel and the so-called unfolded input. Compute individual gradient ℓ₂ norms directly from these matrices (fetched in `get_weight_gradient_factors`) without building up individual gradients explicitly in memory.

Depending on the input dimensions and hyperparameters (out channels, kernel size, ...) of convolution, both approaches perform differently, but I have not done a proper analysis of the trade-offs.

**Comment by SamuelGabriel** [comment]
Hi @f-dangel,

Thanks for your quick reply. I have the same issues I have with my method using batchl2norm. My extension computes the dot product of each gradient with the sum of the gradients and  first I compute the gradient this way and the dot products with it of the single gradients without ever using all the individual gradients, but only applying sequentially the factorization provided by `get_weight_gradient_factors`. 

My thought was that both 1 and 2 are inefficient for some convolutions and I was wondering wether computing batchl2norm (or my method) could be done in a different more efficient way (closer to the convolution) using the unfolded convolution nor computing the per-sample gradients? But I do not have an intuition of how this would work that is why I am asking here. 

I did not think, though, of this trade-off. Thanks for pointing out. I will try to use method 1 on the WideResNet, to see if that maybe works better. Do you know which of the hyper parameters is most important for this trade-off?

**Comment by SamuelGabriel** [comment]
That is what I expected. I also tried to come up with something for some time. Thank You!

**Comment by SamuelGabriel** [comment]
I found a solution that works for computing the dot product between the gradient of each batch item and some other gradient/matrix for convolutions. It boils down to applying the same convolution as in the forward pass, but instead of the original weight use the matrix you want to compute the dot product with as weight and then compute the dot products between the output and the incoming gradient dE/do. I could not find a way, though, that a similar approach can be used to compute l2 norm. 

In code:
torch.einsum('bcwh,bcwh->b',conv(inp1,m),g_o), where m is the weight you wan't to compute the dot product with, inp1 is the input and g_o is the incoming gradient dE/do.

Maybe this inspires you how to do something similar for the l2 norm product, I so far could not do that.

I was also wondering about the implementation of the weight gradient of conv2d _weight_jac_t_mat_prod. First it does not use g_inp, but module.input0 instead, but that seems to be the case for other functions, too. Secondly, I  wondered if it is the most efficient way to always first compute all individual gradients and then sum up, like you do. Do you have any idea, where to find good resources on how this is done in pytorch. I also wondered wether there is some way to just use pytorch's own gradient instead, maybe by adding additional hooks that are run after '.grad's are computed?

**Comment by f-dangel** [comment]
Hey, thanks for the update.

I agree that your workaround for this specific dot product is faster: You want the result of `vᵀ Wᵀ gᵢ` where `v` is a vector, `Wᵀ` is the transposed Jacobian, and `gᵢ` is the incoming gradient for sample `i`.

Instead of applying `Wᵀ` to every `gᵢ` for `i=1, ..., N`, it's better to rewrite the above as `(W v)ᵀ gᵢ`, then apply the Jacobian (note: not the transposed Jacobian) to `v` only once, and then dot with the `gᵢ`. In your case, multiplication by `Wᵀ` was given by `ConvTransposeNd` while application of `W` was a `ConvNd`.

Note that the above does not work anymore when we want the ℓ₂ norms of `Wᵀ gᵢ`, i.e. `gᵢᵀ W Wᵀ gᵢ` for `i=1, ..., N`. There is no way to avoid `N`-fold application of the Jacobians. Still, when we have access to the Jacobian `W`, the ℓ₂ norms can be sped up. We describe this trick for a linear layer on [page 12 of the paper](https://arxiv.org/pdf/1912.10985.pdf). But for convolutions, this matrix is only implicitly represented by the kernel.

Relating to what I said in my [previous reply](https://github.com/f-dangel/backpack/issues/94#issuecomment-678372313), for convolutions, we are computing the ℓ₂ norms as `(gᵢᵀ W) (Wᵀ gᵢ)` for `i=1, ..., N`. Writing code that computes `(gᵢᵀ W Wᵀ gᵢ)` would require low-level code. This should be plenty of work and we're offering it as a [research project](https://uni-tuebingen.de/en/faculties/faculty-of-science/departments/computer-science/lehrstuehle/methods-of-machine-learning/positions/bachelor-master-available-thesis-topics/).

Regarding your question about the output-weight Jacobian for convolutions: Convolution is a linear mapping (in terms of the weight) applied to the input. Hence, the output-weight Jacobian or its transpose (`weight_jac_t`) only depends on the input to the conv layer in a forward pass. If you want to compute the sum of individual gradients, then [PyTorch](https://pytorch.org/) is great. [BackPACK](https://www.backpack.pt) however focuses on other quantities which cannot be computed straightforwardly in [PyTorch](https://pytorch.org/) as they require access to the individual gradients, i.e. before they are summed.

**Comment by SamuelGabriel** [comment]
Thank You for the long answer. I think I now get why there is no way the given torch kernels are enough for the squared l2. Thank You! I actually found a nice way to use pytorch's '.grad' inside your extensions by registering another hook to the tensor I need the grad for and handing it all other information I need over module parameters. I register the hook itself inside the extension. So, actually very similar to how you do things, just for the parameter. 

This project will give a lot of nice student projects I am sure :) Thank You for pushing this, it really helps me with my research and I even by coincidence talked to someone else in the lab who also is depending on you. 👍
